### ISSUE TITLE
Ticks that emits value at start, not end.

### ISSUE THREAD

**Issue by JordanMartinez** [TomasMikula/ReactFX#35]
Coming from TomasMikula/RichTextFX#189, provide additional `ticks` EventStream that emits its values at the start of its interval rather than the end of it.

**Comment by JordanMartinez** [comment]
To make this change, [FxTimer#restart method](https://github.com/TomasMikula/ReactFX/blob/master/reactfx/src/main/java/org/reactfx/util/FxTimer.java#L62-L71) needs to be changed.

To emit values at the end, leave it as it currently is:
````java
timeline.getKeyFrames().setAll(
    new KeyFrame(timeout, ae -> {
        if(seq == expected) {
            action.run();
        }
    })
);
````

To emit values at start, an additional class would need to override restart to:
````java
timeline.getKeyFrames().setAll(
    new KeyFrame(Duration.ZERO, ae -> {
        if(seq == expected) {
            action.run();
        }}), 
    new KeyFrame(timeout)
);
````

One question is how to best handle the static factory methods and class names of FxTimer.
One possible way:
- Rename FxTimer to FxTimerOnEnd
- Add additional class FxTimerOnStart that extends FxTimerOnEnd and overrides the restart method.

Additionally, another question is how to handle the `EventStreams#ticks` method renaming to account for the different types of ticks.

**Comment by TomasMikula** [comment]
I would try to keep just one `FxTimer` class, taking two `Duration` arguments: `period` and `actionTime`. In `restart()`, always add the "action" `KeyFrame` at `actionTime`. If `period` is different from `actionTime`, also add an empty `KeyFrame` at `period`.

I was thinking we could call the new factory method simply `ticks0`, since the first tick occurs at time 0.

**Comment by JordanMartinez** [comment]
> I was thinking we could call the new factory method simply ticks0, since the first tick occurs at time 0.

That's a much simpler/shorter and overall better name than what was coming to my mind. :+1: 

> I would try to keep just one FxTimer class

I agree. So, how about....
````java
public class FxTimer implements Timer {
    // Factory methods that run action when the timer ends
    public static Timer create(java.time.Duration delay, Runnable action) {
        return new FxTimer(delay, delay, action, 1);
    }
    // etc...

    // and their corresponding copies that run action when timer starts
    public static Timer create0(java.time.Duration delay, Runnable action) {
        return new FxTimer(java.time.Duration.ZERO, delay, action, 1);
    }
    // etc...

    // new constructor
    private FxTimer(java.time.Duration actionTime, java.time.Duration period, Runnable action, int cycles) {
        this.actionTime = Duration.millis(actionTime.toMillis());
        this.period = Duration.millis(period.toMillis());
        this.timeline = new Timeline();
        this.action = action;

        timeline.setCycleCount(cycles);
    }

    // new restart method
    @Override
    public void restart() {
        stop();
        long expected = seq;
        if (period == actionTime) {
            timeline.getKeyFrames().setAll(new KeyFrame(actionTime, ae -> {
                if(seq == expected) {
                    action.run();
                }
            }));
        } else {
            timeline.getKeyFrames().setAll(
                new KeyFrame(actionTime, ae -> {
                    if(seq == expected) {
                        action.run();
                    }}),
                new KeyFrame(period)
            );

        }
        timeline.play();
    }
````

As a side note, why is `java.time.Duration` used as opposed to `javafx.util.Duration`?

**Comment by TomasMikula** [comment]
There's some repetition. What about

```java
timeline.getKeyFrames().setAll(new KeyFrame(actionTime, ae -> {
    if(seq == expected) {
        action.run();
    }
}));
if (period != actionTime) {
    timeline.getKeyFrames().add(new KeyFrame(period));
}
```

**Comment by TomasMikula** [comment]
Now, the `if` can be evaluated once in the constructor and not re-evaluated on each restart. So how about this:

```java
    private FxTimer(java.time.Duration actionTime, java.time.Duration period, Runnable action, int cycles) {
        this.actionTime = Duration.millis(actionTime.toMillis());
        this.timeline = new Timeline();
        this.action = action;

        timeline.getKeyFrames().add(new KeyFrame(this.actionTime)); // placeholder for action frame
        if(period != actionTime) { // using != is justified by the factory methods above
            timeline.getKeyFrames().add(new KeyFrame(Duration.millis(period.toMillis())));
        }
        timeline.setCycleCount(cycles);
    }

    @Override
    public void restart() {
        stop();
        long expected = seq;
        timeline.getKeyFrames().set(0, new KeyFrame(actionTime, ae -> {
            if(seq == expected) {
                action.run();
            }
        }));
        timeline.play();
    }
```

**Comment by TomasMikula** [comment]
Resolved by #39.
